### PR TITLE
DTSPO-9695 Refactor pipelines to use CNP templates

### PR DESCRIPTION
**Pull request by vijay-tenoy**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/DTSPO-9695

### Change description ###
Refactor pipelines to use CNP templates

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: azure-pipelines.yml [hunk 1]
```diff
@@ -1,20 +1,28 @@
-name: HMCTS CFT AKS Azure Infrastructure Deployment Pipeline
+name: "${{ parameters.action }} - ${{ parameters.env }}.${{ parameters.cluster }}.${{ parameters.location }}"
```
Comment (thomast1906) [inline azure-pipelines.yml:1]: very nice addition

File: azure-pipelines.yml [hunk 2]
```diff
@@ -1,20 +1,28 @@
...
 trigger: none
 
 pool:
   vmImage: 'ubuntu-latest'
 
+resources:
+  repositories:
+    - repository: cnp-azuredevops-libraries
+      type: github
+      ref: refs/heads/DTSPO-9695-make-tf-varfile-configurable
```
Comment (thomast1906) [inline azure-pipelines.yml:12]: can we ensure this branch has been tested against other pipelines

**Comment by thomast1906** [comment]
Pipeline for AAT https://dev.azure.com/hmcts/CNP/_build/results?buildId=286392&view=results

Pipeline for SBOX https://dev.azure.com/hmcts/CNP/_build/results?buildId=286366&view=results

File: azure-pipelines.yml [hunk 3]
```diff
@@ -100,64 +130,86 @@ stages:
...
+              overrideAction: ${{ parameters.action }}
               environment: ${{ parameters.env }}
+              component: 'network-rg'
               location: ${{ parameters.location }}
-              stack: 'network-rg'
-              project: $(project)
-              tfversion: $(tfversion)
-              tfInitSub: ${{ variables.tfInitSub }}
-              builtFrom: $(Build.Repository.Name)
+              serviceConnection: $(serviceConnection)
```
Comment (timja) [inline azure-pipelines.yml:152]: doesn't this stage need to use the `OPS-APPROVAL-GATE-MGMT-ENVS-CNP` service connection? (to do network peering against any subscription)

File: pipeline-steps/destroy-service.yaml [hunk 4]
```diff
@@ -8,13 +8,11 @@ parameters:
   targetCommand: ''
   builtFrom: ''
   product: ''
+  serviceConnection: ''
```
Comment (timja) [inline pipeline-steps/destroy-service.yaml:11]: I'm pretty sure destroy-service.yaml is not actually used in practice and could be removed in a separate PR (non-blocking)

File: pipeline-steps/keyvault-read.yaml [hunk 5]
```diff
@@ -5,49 +5,20 @@ parameters:
...
+  - task: AzureKeyVault@1
+    displayName: 'Get Secrets from KeyVault'
+    inputs:
+      azureSubscription: ${{ parameters.serviceConnection }}
+      keyVaultName: $(controlKeyVault)
+      secretsFilter: 'sp-application-id,sp-token'
+
+  - script: |
+      echo '##vso[task.setvariable variable=ARM_CLIENT_ID]$(sp-application-id)'
+      echo '##vso[task.setvariable variable=ARM_CLIENT_SECRET]$(sp-token)'
```
Comment (timja) [inline pipeline-steps/keyvault-read.yaml:22]: future enhancement likely, I'm pretty sure these aren't used at all, could be cleaned up separately

File: pipeline-steps/deploy-network.yaml [hunk 6]
```diff
@@ -5,71 +5,27 @@ parameters:
   project: ''
   tfversion: ''
   tfvars: ''
-  terraformServiceConnection: ''
+  serviceConnection: ''
```
Comment (timja) [inline pipeline-steps/deploy-network.yaml:8]: it seems like this file isn't used anymore looking at the pipeline?

File: pipeline-steps/genesis.yaml [hunk 7]
```diff
@@ -4,72 +4,26 @@ parameters:
   serviceConnection: ''
   stack: ''
   project: ''
-  tfversion: ''
-  tfInitSub: ''
-  builtFrom: ''
   product: ''
+  tfInitSub: ''
```
Comment (timja) [inline pipeline-steps/genesis.yaml:8]: is this file (genesis.yaml) actually required anymore?

i.e what's stopping it from using the `steps/terraform.yaml@cnp-azuredevops-libraries` ?

File: pipeline-steps/keyvault-read.yaml (hunk 5 above)
Comment (thomast1906) [reply pipeline-steps/keyvault-read.yaml:22]: I dont think they are used at all - @vijay-tenoy can you run a plan with these commented out?

Comment (vijay-tenoy) [reply pipeline-steps/keyvault-read.yaml:22]: Sure, will cleanup and report back here

File: pipeline-steps/genesis.yaml (hunk 7 above)
Comment (vijay-tenoy) [reply pipeline-steps/genesis.yaml:8]: There is just one extra step in genesis.yaml (pasted below) before teraform plan and apply. Maybe it could be moved to the main pipeline?

`- script: rm -fr .terraform/modules/${{ parameters.stack }}`

Comment (vijay-tenoy) [reply pipeline-steps/genesis.yaml:8]: Note: the genesis.yaml file already uses steps/terraform.yaml@cnp-azuredevops-libraries, just has the extra step pasted in previous comment

Comment (timja) [reply pipeline-steps/genesis.yaml:8]: pretty sure not needed, no one has touched that file in a long time i would just remove it and see if any issues

File: pipeline-steps/deploy-network.yaml (hunk 6 above)
Comment (vijay-tenoy) [reply pipeline-steps/deploy-network.yaml]: Correct. Removed the file and tested the pipeline.

File: pipeline-steps/keyvault-read.yaml (hunk 5 above)
Comment (vijay-tenoy) [reply pipeline-steps/keyvault-read.yaml]: Removed  keyvault-read.yaml entirely and directly used `steps/keyvault-read.yaml@cnp-azuredevops-libraries` in templates that needed to read from keyvault.

File: pipeline-steps/destroy-service.yaml (hunk 4 above)
Comment (vijay-tenoy) [reply pipeline-steps/destroy-service.yaml]: Removed the destroy yaml file and the associated state in azure pipelines. Deleting code is easy!

File: azure-pipelines.yml (hunk 3 above)
Comment (vijay-tenoy) [reply azure-pipelines.yml:151]: Correct. Changed the network stage to use the right service connection i.e. `OPS-APPROVAL-GATE-MGMT-ENVS-CNP`

File: pipeline-steps/keyvault-read.yaml (hunk 5 above)
Comment (timja) [reply pipeline-steps/keyvault-read.yaml]: cool that one looks more sensible:
https://github.com/hmcts/cnp-azuredevops-libraries/blob/master/steps/keyvault-read.yaml

it just sets useful vars like sub id, tenant id, vault name and storage account

File: pipeline-steps/managed_identity.yaml [hunk 8]
```diff
@@ -11,7 +11,7 @@ parameters:
 
 steps:
 
-  - template: keyvault-read.yaml
+  - template: steps/keyvault-read.yaml@cnp-azuredevops-libraries
```
Comment (timja) [inline pipeline-steps/managed_identity.yaml:14]: is this file managed_identity.yaml still needed?

File: azure-pipelines.yml (hunk 2 above)
Comment (vijay-tenoy) [reply azure-pipelines.yml:12]: Pipeline for `azure-private-dns` has been tested using the new common pipeline template changes https://dev.azure.com/hmcts/CNP/_build/results?buildId=279519&view=results

I'll test a few more pipelines just to be on the safe side

File: pipeline-steps/managed_identity.yaml (hunk 8 above)
Comment (vijay-tenoy) [reply pipeline-steps/managed_identity.yaml:14]: Not currently being used. We directly use `steps/terraform.yaml@cnp-azuredevops-libraries` in the Managed identity stage. Removed the file

File: pipeline-steps/genesis.yaml (hunk 7 above)
Comment (vijay-tenoy) [reply pipeline-steps/genesis.yaml]: Done. Removed `genesis.yaml` and used `steps/terraform.yaml@cnp-azuredevops-libraries` directly without the extra step mentioned above. No issues in pipeline

**Comment by timja** [comment]
conflicted with master

File: azure-pipelines.yml [hunk 9]
```diff
@@ -90,74 +119,125 @@ stages:
...
+                -var project=$(project)
+
       - job: DeployInfrastructure
         dependsOn: NetworkRg
         steps:
-          - template: pipeline-steps/deploy-network.yaml
+          - template: steps/set-build-repo-suffix-env-var.yaml@cnp-azuredevops-libraries
+
+          - task: AzureCLI@2
+            displayName: Determine Key Vault and Storage names
```
Comment (timja) [inline azure-pipelines.yml:170]: why did you need to copy this, you should just be able to use https://github.com/hmcts/cnp-azuredevops-libraries/blob/master/steps/keyvault-read.yaml
